lib/semantics/check-return.cc

```diff
@@ -33,18 +30,17 @@ void ReturnStmtChecker::Leave(const parser::ReturnStmt &returnStmt) {
   // C1575 The scalar-int-expr is allowed only in the inclusive scope of a
   // subroutine subprogram.
   const auto &scope{context_.FindScope(context_.location().value())};
-  const auto *subprogramScope{FindContainingSubprogram(scope)};
-  if (!subprogramScope) {
-    context_.Say(
-        "RETURN must in the inclusive scope of a SUBPROGRAM"_err_en_US);
-    return;
-  }
-  if (returnStmt.v && subprogramScope->kind() == Scope::Kind::Subprogram) {
-    if (IsFunction(*subprogramScope->GetSymbol())) {
-      context_.Say(
-          "RETURN with expression is only allowed in SUBROUTINE subprogram"_err_en_US);
+  if (const auto *subprogramScope{FindContainingSubprogram(scope)}) {
+    if (returnStmt.v && subprogramScope->kind() == Scope::Kind::Subprogram) {
+      if (IsFunction(*subprogramScope->GetSymbol())) {
+        context_.Say(
+            "RETURN with expression is only allowed in SUBROUTINE subprogram"_err_en_US);
+      }
     }
+  } else {
+    context_.Say(
+        "RETURN must be in the inclusive scope of a subprogram"_err_en_US);
```

Comment:
I didn't see a test case for this condition.

Reply:
There should have already been one, but I couldn't find it, so I'll add one.

Reply:
Or I'll delete the message, since it could only be produced in a program already eliciting parsing errors about an executable part in a `MODULE`, `SUBMODULE`, or `BLOCK DATA` program unit.

Reply:
What about in a `PROGRAM`? Are we allowing that as an extension?

Reply:
GNU Fortran, PGI, and XLF accept it.  Intel Fortran and NAG do not.  So we probably should.